src/utils.cpp

```diff
@@ -0,0 +1,48 @@
+#include "utils.h"
+
+#include <Preferences.h>
+#include <esp_random.h>
+
+// RTC persistent data for PIN cache (survives deep sleep)
+RTC_DATA_ATTR uint32_t rtc_cachedPIN = 0;
+
+// NVS storage for PIN (survives power cycles)
+static Preferences prefs;
+
+namespace Utils
+{
+
+void initPIN()
+{
+  prefs.begin("zivyobraz", false);
+
+  if (!prefs.isKey("pin"))
+  {
+    // Generate 8-digit PIN without leading zeros (range: 10000000 - 99999999)
+    rtc_cachedPIN = (esp_random() % 90000000) + 10000000;
+    prefs.putULong("pin", rtc_cachedPIN);
+    Serial.print("Generated new device PIN: ");
+    Serial.println(rtc_cachedPIN);
+  }
+  else
+  {
+    rtc_cachedPIN = prefs.getULong("pin", 0);
```

Comment:
Same as above
```suggestion
    rtc_cachedPIN = prefs.getULong("pin");
```

// multiple places